public/pages/AnomalyCharts/utils/anomalyChartUtils.ts

```diff
@@ -0,0 +1,179 @@
+          value_count: {
+            field: 'monitor_id',
+          },
+        },
+      },
+      query: {
+        bool: {
+          filter: [
+            {
+              term: {
+                monitor_id: monitorId,
+              },
+            },
+            {
+              range: {
+                start_time: {
+                  gte: startTime,
+                  format: 'epoch_millis',
+                },
+              },
+            },
+          ],
+        },
+      },
+      sort: [{ start_time: { order: 'desc' } }],
+    },
+  };
+};
+
+export const convertAlerts = (response: any): MonitorAlert[] => {
+  const hits = get(response, 'data.response.hits.hits', []);
+  return hits.map((alert: any) => {
+    return {
+      monitorName: get(alert, '_source.monitor_name'),
+      triggerName: get(alert, '_source.trigger_name'),
+      severity: get(alert, '_source.severity'),
+      state: get(alert, '_source.state'),
+      error: get(alert, '_source.error_message'),
+      startTime: get(alert, '_source.start_time'),
+      endTime: get(alert, '_source.end_time'),
+      acknowledgedTime: get(alert, '_source.acknowledged_time'),
+    };
+  });
+};
+
+const getAlertMessage = (alert: MonitorAlert): string => {
+  const message = alert.endTime
+    ? `There is a severity ${alert.severity} alert with state ${
+        alert.state
+      } between ${dateFormatter(alert.startTime)} and ${dateFormatter(
+        alert.endTime
+      )}.`
+    : `There is a severity ${alert.severity} alert with state ${
+        alert.state
+      } from ${dateFormatter(alert.startTime)}.`;
+  return alert.error ? `${message} Error message: ${alert.error}` : message;
+};
+
+export const generateAlertAnnotations = (alerts: MonitorAlert[]): any[] => {
+  return alerts.map((alert: MonitorAlert) => ({
+    header: dateFormatter(alert.startTime),
+    dataValue: alert.startTime,
+    details: getAlertMessage(alert),
+  }));
+};
+
+const findLatestAnomaly = (anomalies: any[]) => {
+  let latestAnomaly = anomalies[0];
+  for (let i = 1, len = anomalies.length; i < len; i++) {
+    let item = anomalies[i];
+    latestAnomaly =
+      item.startTime > latestAnomaly.startTime ? item : latestAnomaly;
+  }
+  return latestAnomaly;
+};
+
+export const getAnomalySummary = (anomalies: any[]): AnomalySummary => {
+  let minConfidence = 1.0,
+    maxConfidence = 0.0;
+  let minAnomalyGrade = 1.0,
+    maxAnomalyGrade = 0.0;
+  const targetAnomalies = anomalies.filter(anomaly => {
+    if (anomaly.anomalyGrade < minAnomalyGrade) {
+      minAnomalyGrade = anomaly.anomalyGrade;
+    }
+    if (anomaly.anomalyGrade > maxAnomalyGrade) {
+      maxAnomalyGrade = anomaly.anomalyGrade;
+    }
+    if (anomaly.confidence < minConfidence) {
+      minConfidence = anomaly.confidence;
+    }
+    if (anomaly.confidence > maxConfidence) {
+      maxConfidence = anomaly.confidence;
+    }
+    if (anomaly.anomalyGrade > 0) {
+      return true;
+    }
+  });
```

Comment:
To me, this seems kind of odd updating the max/min confidence/grade inside of `filter()`, since it seems we're only filtering by anomalies having a grade > 0. Maybe have separate calls for each var, like `const minConfidence = Math.min(...anomalies.map(anomaly => anomaly.anomalyGrade), 1.0)`. Then the `filter()` call could be simpler. Although this would increase space complexity. Thoughts?